applications/ParticleMechanicsApplication/custom_utilities/mpm_particle_generator_utility.h

```diff
@@ -50,11 +50,253 @@ namespace MPMParticleGeneratorUtility
      * @brief Construct material points or particles from given initial mesh
      * @details Generating particles using a designated shape functions
      */
+    template<SizeType TDimension>
     void GenerateMaterialPointElement(  ModelPart& rBackgroundGridModelPart,
                                         ModelPart& rInitialModelPart,
                                         ModelPart& rMPMModelPart,
                                         bool IsAxisSymmetry = false,
-                                        bool IsMixedFormulation = false);
+                                        bool IsMixedFormulation = false) {
+        // Initialize zero the variables needed
+        std::vector<array_1d<double, 3>> xg = { ZeroVector(3) };
+        std::vector<array_1d<double, 3>> mp_displacement = { ZeroVector(3) };
+        std::vector<array_1d<double, 3>> mp_velocity = { ZeroVector(3) };
+        std::vector<array_1d<double, 3>> mp_acceleration = { ZeroVector(3) };
+        std::vector<array_1d<double, 3>> mp_volume_acceleration = { ZeroVector(3) };
+
+        std::vector<Vector> mp_cauchy_stress_vector = { ZeroVector(6) };
+        std::vector<Vector> mp_almansi_strain_vector = { ZeroVector(6) };
+        std::vector<double> mp_pressure = { 0.0 };
+
+        std::vector<double> mp_mass(1);
+        std::vector<double> mp_volume(1);
+
+        // Determine element index: This convention is done in order for the purpose of visualization in GiD
+        const unsigned int number_elements = rBackgroundGridModelPart.NumberOfElements() + rInitialModelPart.NumberOfElements();
+        const unsigned int number_nodes = rBackgroundGridModelPart.NumberOfNodes();
+        unsigned int last_element_id = (number_nodes > number_elements) ? (number_nodes + 1) : (number_elements + 1);
+
+        BinBasedFastPointLocator<TDimension> SearchStructure(rBackgroundGridModelPart);
+        SearchStructure.UpdateSearchDatabase();
+        typename BinBasedFastPointLocator<TDimension>::ResultContainerType results(100);
+
+        // Loop over the submodelpart of rInitialModelPart
+        for (ModelPart::SubModelPartIterator submodelpart_it = rInitialModelPart.SubModelPartsBegin();
```

Comment:
just to tell that this loop will become easier if you use the range based for

~~~
for(auto& submodelpart : rInitialModelPart.SubModelParts())
{...}
~~~